### PR TITLE
improving request-promise-native: expanding defs and linting

### PR DESCRIPTION
**Pull request by jcgillespie**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: 
[request-promise defs](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/request-promise/index.d.ts)
[request-promise docs](https://github.com/request/request-promise)
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dslint/dt.json" }`.

### CODE REVIEW

File: types/request-promise-native/index.d.ts [hunk 1]
```diff
@@ -6,13 +6,15 @@
...
 
     namespace requestPromise {
         interface RequestPromise extends request.Request {
             then<TResult>(onfulfilled?: (value: any) => TResult | PromiseLike<TResult>, onrejected?: (reason: any) => TResult | PromiseLike<TResult> | void): Promise<TResult>;
-
             catch(onrejected?: (reason: any) => any | PromiseLike<any> | void): Promise<any>;
-
+            catch(type: errors.RequestErrorConstructor, onrejected?: (reason: errors.RequestError) => void): Promise<any>;
+            catch(type: errors.StatusCodeErrorConstructor, onrejected?: (reason: errors.StatusCodeError) => void): Promise<any>;
+            catch(type: errors.TransformErrorConstructor, onrejected?: (reason: errors.TransformError) => void): Promise<any>;
```
Comment (gustavohenke) [inline types/request-promise-native/index.d.ts]: These should definitely be dropped, as `request-promise-native` uses native promises - and these don't have error type selection, like Bluebird promises do (used by `request-promise` itself).

**Review by gustavohenke — CHANGES_REQUESTED** [review]

File: types/request-promise-native/index.d.ts (hunk 1 above)
Comment (jcgillespie) [inline types/request-promise-native/index.d.ts]: Will do. Thanks!

Comment (jcgillespie) [inline types/request-promise-native/index.d.ts]: @gustavohenke that update is committed/built now.
Thanks again!

**Comment by jcgillespie** [comment]
@gustavohenke thoughts on the latest?

File: types/request-promise-native/index.d.ts [hunk 2]
```diff
@@ -1,31 +1,75 @@
-// Type definitions for request-promise-native v1.0.3
+// Type definitions for request-promise-native 1.1
```
Comment (gustavohenke) [inline types/request-promise-native/index.d.ts]: There is no v1.1 of `request-promise-native`: https://github.com/request/request-promise-native/releases

File: types/request-promise-native/index.d.ts [hunk 3]
```diff
@@ -1,31 +1,75 @@
...
-    export type OptionsWithUri = request.UriOptions & RequestPromiseOptions;
-    export type OptionsWithUrl = request.UrlOptions & RequestPromiseOptions;
-    export type Options = OptionsWithUri | OptionsWithUrl;
+    let requestPromise: request.RequestAPI<requestPromise.RequestPromise, requestPromise.RequestPromiseOptions, request.RequiredUriUrl>;
+    export = requestPromise;
 }
 
-declare var requestPromise: request.RequestAPI<requestPromise.RequestPromise, requestPromise.RequestPromiseOptions, request.RequiredUriUrl>;
-export = requestPromise;
+declare module 'request-promise-native/errors' {
```
Comment (gustavohenke) [inline types/request-promise-native/index.d.ts]: Unused module

File: types/request-promise-native/request-promise-native-tests.ts [hunk 4]
```diff
@@ -1,65 +1,69 @@
 import * as rp from 'request-promise-native';
+import * as errors from 'request-promise-native/errors';
```
Comment (gustavohenke) [inline types/request-promise-native/request-promise-native-tests.ts]: Unused import

**Review by gustavohenke — CHANGES_REQUESTED** [review]
I'm quite out of the TypeScript world lately, so I may be lacking to see any further changes than linting fixes, after you removed the error type selection.
Is that correct?

If that's the case, just a tip: in my own projects I prefer to rebase the branch and remove the isolated commits instead of adding more commits to undo stuff

File: types/request-promise-native/index.d.ts (hunk 2 above)
Comment (jcgillespie) [inline types/request-promise-native/index.d.ts]: Correct. The version is for the types, not the library. The linter disallows the "v" prefix and the "patch" number. Didn't think that going "back" to a `1.0` was the right direction, so I went to `1.1

File: types/request-promise-native/index.d.ts (hunk 3 above)
Comment (jcgillespie) [inline types/request-promise-native/index.d.ts]: Matched the `request-promise` type definitions.

File: types/request-promise-native/request-promise-native-tests.ts (hunk 4 above)
Comment (jcgillespie) [inline types/request-promise-native/request-promise-native-tests.ts]: Matched the `request-promise` type definitions.

**Comment by jcgillespie** [comment]
@gustavohenke The impetus for the PR was adding the `/errors` module, which did not exist in the native type definitions, but does exist in both [request-promise types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/request-promise/index.d.ts) and in the underlying [promise-core library](https://github.com/request/promise-core/blob/master/lib/errors.js)

Everything else was driven by one of two things.
1. The linter corrections
2. matching the `request-promise` type definitions.

I didn't anticipate making additional changes, but in walking through the contribution checklist I realized the definitions didn't adhere to the current guidelines. I can abandon this PR and re-open one that is rebased if you prefer.

**Comment by jcgillespie** [comment]
I'll drop this one and resubmit just the piece I need.

**Comment by gustavohenke** [comment]
@jcgillespie sorry for taking this long to answer you 🙈 
I just reviewed your other PR #16216.

**Comment by jcgillespie** [comment]
No worries. I broke the lint changes out into another set of commits that I'll submit separately.

**Pull request closed without merging**